### DISCUSSION TITLE
How does AsyncSession handle CancelledError with operations in flight?

### DISCUSSION THREAD

**Discussion by lunalucadou** [sqlalchemy/sqlalchemy#12119] [Usage Questions] [answered]
I've been working on converting a method to async since we have to update a value in several systems and, since they have no dependency on each other, doing them simultaneously could help speed things up.

This is the existing code:

```python
# OurDbService
def create_update_name(self, name):
  with Session(self.db_engine) as s:
    new_name = s.merge(Name.from_dict(name))
    s.commit()
    s.refresh(new_name)
    return new_name

# UpstreamDbService
NAME_UPDATE_CALL = text("CALL UPSERT_NAME(FIRST_NAME => :first_name, ... )")
def create_update_name(self, name):
  with self.upstream_engine.begin() as conn:
    conn.execute(NAME_UPDATE_CALL, { "first_name": name["first_name"], ...: ...})

# NameService
def create_name(self, request):
  requested_name = validate_and_extract_name_to_dict(request)
  self.upstream_db.create_update_name(requested_name)
  saved_name = self.our_db.create_update_name(requested_name)
  # ...
```

And this is what I've come up with for an async implementation:

```python
# OurDbService
async def async_create_update_name(self, name):
  async with AsyncSession(self.async_db_engine) as s:
    new_name = await s.merge(Name.from_dict(name))
    await s.commit()
    await s.refresh(new_name)
    return new_name

# UpstreamDbService
async def async_create_update_name(self, name):
  # OracleDB thick client doesn't seem to support true async so this may be the best I can do
  # https://python-oracledb.readthedocs.io/en/latest/user_guide/asyncio.html#concurrent-programming-with-asyncio
  with self.upstream_engine.begin() as conn:
    conn.execute(NAME_UPDATE_CALL, { "first_name": name["first_name"], ...: ...})

# NameService
async def async_create_name(self, request):
  requested_name = validate_and_extract_name_to_dict(request)
  saved_name = None
  try:
    async with asyncio.timeout(10):
      upstream_db_task = asyncio.create_task(self.upstream_db.async_create_update_name(requested_name))
      our_db_task = asyncio.create_task(self.our_db.async_create_update_name(requested_name))
      saved_name = await our_db_task
  except asyncio.CancelledError:
    if saved_name:
      self.our_db.delete_name(saved_name)
  # ...
```

The reason for the `asyncio.timeout(10)` is because we have 15 seconds to complete everything before the reserve proxy closes the connection, and because upstream timeouts rarely time out fast, we wouldn't have time to clean up our local changes if done asynchronously (when done synchronously, we just never got to the point of committing the changes locally to begin with when upstream timed out).

Since we re-poll upstream periodically, if the stored procedure succeeds but we get killed before finding out, we will pick up the change within several hours, but since upstream does not poll us, if we only make a change in our database, it will be wiped out within several hours (so it's preferable to report a false negative to our end users than a false positive).

I suspect that, 99% of the time, the timeout and undo logic I built will be sufficient. But while writing tests, I began to wonder: what happens if there is a slowdown (in the remote call, local DB update, or just randomly between function calls) such that the timer runs out while our session is still open? That is, if the timer expires when:

* `await s.commit()` has been called, and the DB received it, but our application is waiting on the response/hasn't processed the response
* `await s.commit()` has finished and our application knows about it, but the application is either waiting on `await s.refresh()` or has completed all that and the return statement hasn't been processed yet

Could this leave us in an inconsistent state (changes only reflected locally), or does the AsyncSession auto-rollback when it encounters an exception?

If sessions do not have any auto-rollback, would this be a good solution:

```python
# OurDbService
async def async_create_update_name(self, name):
  async with AsyncSession(self.async_db_engine) as s:
    try:
      new_name = await s.merge(Name.from_dict(name))
      await s.commit()
      await s.refresh(new_name)
      return new_name
    except asyncio.CancelledError:
      await s.rollback()
      raise
```

or is there a better/more official way, perhaps by configuring the session to automatically handle these exceptions?

**Comment by zzzeek** [comment]
> update a value in several systems

what's "several"?   fewer than 20?   you can just use [concurrent.futures](https://docs.python.org/3/library/concurrent.futures.html) with a threadpool and save a lot of headache

> `self.async_db_engine`
> `# OracleDB thick client `
> `with self.upstream_engine.begin() as conn:`

we would need to know a lot more about what these different engines are and what drivers are involved.   if you are using a blocking Oracle driver, asyncio gets you absolutely nothing.   `concurrent.futures` gets you real concurrency fully supported by any driver.   multiple engines of different types in play here looks very exotic

>  `async with asyncio.timeout(10):` `asyncio.create_task(self.upstream_db.async_create_update_name(requested_name))

your `async_create_update_name` is calling blocking IO.   nothing async will happen here, not timeouts, nothing else.....

overall this looks to be working against how asyncio is supposed to be used and especially with Oracle drivers and legacy code, unless you have very novel use cases for many hundreds of concurrent DB activities (and even then...) you want to just use basic thread-based concurrency to get started.

**Reply by lunalucadou** [reply]
For now it's 3 places we have to update at one time, but if other upstream systems were to request similar functionality, this could balloon much further with an absolute worst case being a combination of 3-16 systems, depending on who was being updated and who sent the request (thankfully, it seems unlikely we will ever have to write to more than 4 at a time).

It's Oracle that seems to pose the biggest challenge, both because that's the primary source of timeouts and because we have to use the thick client, which appears to have no true async support. The other two (MySQL and Elasticsearch) have full async support.

> you want to just use basic thread-based concurrency to get started

I had assumed threading would be best for more intensive use cases (like if we did need to update in 16 places), especially given that it involves more overhead than async.

Hypothetically, if we could use the Oracle thin client with its full async support, would something like this be a good usage of asyncio? (I'm adding in the Elastic updates as well so you can better visualize the flow. The reason Elastic depends on our local DB update completing is because we replicate the auto-incremented IDs, and if a name is being created, we won't know the ID until it's committed. We're considering adding UUIDs so we don't have to process someone in SQL before creating their Elastic documents, but that's neither here nor there for now.)

```python
# OurDbService
async def async_create_update_name(self, name):
  async with AsyncSession(self.async_db_engine) as s:
    new_name = await s.merge(Name.from_dict(name))
    await s.commit()
    await s.refresh(new_name)
    return new_name

# OurElasticService
async def async_create_update_name(self, sql_name, request):
  base_person = await AsyncDocument.get(id=sql_name.person.id, index=request.caller_index, using=self.async_connection)
  elastic_name = EsName.from_sql_name(sql_name)
  # add or replace name on person ...
  asyncio.create_task(base_person.save())
  return elastic_name

# UpstreamDbService
async def async_create_update_name(self, dict_name):
  async with self.async_upstream_engine.begin() as conn:
   asyncio.create_task(conn.execute(NAME_UPDATE_CALL, { "first_name": dict_name["first_name"], ...: ...}))

# NameService
async def async_create_name(self, request):
  requested_name = validate_and_extract_name_to_dict(request)
  saved_name = None
  try:
    async with asyncio.timeout(10):
      upstream_db_task = asyncio.create_task(self.upstream_db.async_create_update_name(requested_name))
      our_db_task = asyncio.create_task(self.our_db.async_create_update_name(requested_name))
      saved_name = await our_db_task
      our_es_task = asyncio.create_task(self.our_es.async_create_update_name(saved_name, request))
  except asyncio.CancelledError:
    if saved_name:
      self.our_db.delete_name(saved_name)
      self.our_es.delete_name(saved_name, request)
  # ...
```

And if that's a bad use for asyncio, would that be because I'm approaching async wrong, or because this use case is just better suited to threading to begin with? (I've only ever done toy examples of async stuff before, nothing involving a proper dependency chain. I have a decent amount of C/C++ threading experience, but none in Python.)

Alternatively, if the approach is fine but Oracle being blocking is the problem, would it be smarter to go all-in on threading or a hybrid approach of async + a separate thread for Oracle?

**Reply by CaselIT** [reply, marked as answer]
Also in this example you are not awaiting many tasks.

> I had assumed threading would be best for more intensive use cases (like if we did need to update in 16 places), especially given that it involves more overhead than async

If you use a thread pool that's standard when the application starts the overhead is likely insignificant with asyncio. Considering that you are mixing asyncio with the sync oracle db it's likely a lot less.

Note that while you are waiting to oracle no other async work will happen, since that's blocking

**Reply by lunalucadou** [reply]
> If you use a thread pool that's standard when the application starts the overhead is likely insignificant with asyncio

That's good to hear. Startup overhead would be our primary concern in doing async vs threaded (assuming we could do async, of course).

**Comment by CaselIT** [comment]
Just a note: in your example you seem not to wait the task `upstream_db_task`

**Reply by lunalucadou** [reply]
Unless I'm misunderstanding the documentation, `asyncio.create_task` auto-starts the task. You only need to wait for a task if there's some dependency, such as acting on the returned value.

(Technically, you do need to store a reference to it to prevent it from getting garbage collected, but I left that code out since it's superfluous for the snippet.)

**Reply by CaselIT** [reply]
sure but if you don't you don't know if it ended of if an error was raised there

**Reply by lunalucadou** [reply]
I had attached a callback with error checking, but I cut that from the sample code for brevity.
